File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        int i = nums.length - 1, low = 0, high = i;
        int[] res = new int[i + 1];
        while(low <= high) {
            int lo = nums[low] * nums[low];
            int hi = nums[high] * nums[high];
            if(lo <= hi) {
                res[i] = hi;
                --high;
            } else {
                res[i] = lo;
                ++low;
            }
            --i;
        }
        return res;
    }
}

File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int i = nums.size() - 1, l = 0, h = i;
        vector<int> res(i + 1);
        while(l <= h) {
            int lo = square(nums[l]);
            int hi = square(nums[h]);
            if(lo <= hi) {
                res[i] = hi;
                --h;
            }
            else {
                res[i] = lo;
                ++l;
            }
            --i;
        }
        return res;
    }
    
    int square(int x) {
        return x * x;
    }
};